notebook/notebookapp.py

```diff
@@ -2289,14 +2319,16 @@ def list_running_servers(runtime_dir=None):
     if not os.path.isdir(runtime_dir):
         return
 
+    api_version_json_bytes = json.dumps({"version":__version__}).encode()
     for file_name in os.listdir(runtime_dir):
         if re.match('nbserver-(.+).json', file_name):
             with io.open(os.path.join(runtime_dir, file_name), encoding='utf-8') as f:
                 info = json.load(f)
 
-            # Simple check whether that process is really still running
+            # Actively check whether that process is really available via an HTTP request
             # Also remove leftover files from IPython 2.x without a pid field
-            if ('pid' in info) and check_pid(info['pid']):
+            response = kernel_request(info, path=url_path_join(info.get('base_url') or '/','/api'))
```

Comment:
As noted previously, `try/except` should surround this call or have the method handle it.  

Since you might want to distinguish two different behaviors for `GET 'api'` and `POST 'api/shutdown'` and will likely need to call `GET 'api'` multiple times (per my previous comment), I would suggest a `ping_server()` method that just takes the `server_info` and issues the `GET` request, returning `True` or `False` to indicate the server is running or not.

Reply:
first stage, i change the core of list_running_servers() to reliable live HTTP request; and keep their behaviors the same;
and previous check_pid() let caller make decisions to exceptions;

suggestion of a ping_server() implemented.